packages/node/warpsync/syncing.go

```diff
@@ -0,0 +1,384 @@
+
+	epochBlocks := make(map[epoch.Index][]*blockReceived)
+	tangleRoots := make(map[epoch.Index]*smt.SparseMerkleTree)
+	epochSyncEnds := make(map[epoch.Index]*epochSyncEnd)
+
+	// We do not request the start nor the ending epoch, as we know the beginning and the end of the chain.
+	for ei := endEpoch - 1; ei > startEpoch; ei-- {
+		db, _ := database.NewMemDB()
+		tangleRoots[ei] = smt.NewSparseMerkleTree(db.NewStore(), db.NewStore(), lo.PanicOnErr(blake2b.New256(nil)))
+		m.requestEpochBlocks(ei, epoch.ComputeEC(ecChain[ei]))
+	}
+
+	// Counters to check termination conditions.
+	epochsToReceive := endEpoch - startEpoch - 1
+	epochBlocksLeft := make(map[epoch.Index]int64)
+
+	m.log.Debug(">> START")
+
+startReadLoop:
+	for {
+		select {
+		case epochSyncStart := <-m.epochSyncStartChan:
+			ei := epochSyncStart.ei
+			ec := epochSyncStart.ec
+			if ei <= startEpoch || ei > endEpoch-1 {
+				m.log.Warnf("received epoch starter for epoch %d while we are syncing from %d to %d", ei, startEpoch, endEpoch)
+				continue
+			}
+			if ec != epoch.ComputeEC(ecChain[ei]) {
+				m.log.Warnf("received epoch starter on wrong EC chain")
+				continue
+			}
+			if _, exists := epochBlocksLeft[ei]; exists {
+				m.log.Warnf("received starter for already-started epoch %d", ei)
+				continue
+			}
+
+			epochBlocksLeft[ei] = epochSyncStart.blocksCount
+			m.log.Debugw("read epoch block count", "EI", ei, "blockCount", epochBlocksLeft[ei])
+
+			epochsToReceive--
+			m.log.Debugf("epochs left %d", epochsToReceive)
+			if epochsToReceive == 0 {
+				break startReadLoop
+			}
+		case <-ctx.Done():
+			return fmt.Errorf("cancelled while syncing epoch range %d to %d: %s", startEpoch, endEpoch, ctx.Err())
+		}
+	}
+
+	epochsToReceive = 0
+	for _, epochBlockCount := range epochBlocksLeft {
+		if epochBlockCount > 0 {
+			epochsToReceive++
+		}
+	}
+	epochReceivedBlocks := make(map[epoch.Index]map[tangle.BlockID]types.Empty)
+
+	m.log.Debug(">> BLOCKS")
+
+	for {
+		if epochsToReceive == 0 {
+			break
+		}
+		select {
+		case epochSyncBlock := <-m.epochSyncBatchBlockChan:
+			ei := epochSyncBlock.ei
+			if ei <= startEpoch || ei > endEpoch-1 {
+				m.log.Warnf("received block for epoch %d while we are syncing from %d to %d", ei, startEpoch, endEpoch)
+				continue
+			}
+			if epochSyncBlock.ec != epoch.ComputeEC(ecChain[ei]) {
+				m.log.Warnf("received block on wrong EC chain")
+				continue
+			}
+			if _, exists := epochSyncEnds[ei]; exists {
+				m.log.Warnf("received block for already-ended epoch %d", ei)
+				continue
+			}
+			if epochBlocks[ei] == nil {
+				epochBlocks[ei] = make([]*blockReceived, 0)
+				epochReceivedBlocks[ei] = make(map[tangle.BlockID]types.Empty)
+			}
+
+			block := epochSyncBlock.block
+			if _, exists := epochReceivedBlocks[ei][block.ID()]; exists {
+				continue
+			}
+
+			m.log.Debugw("read block", "peer", epochSyncBlock.peer.ID(), "EI", ei, "blockID", block.ID())
+
+			tangleRoots[ei].Update(block.IDBytes(), block.IDBytes())
```

Comment:
iiuc: as it is currently where we request the epochs from all neighbors there's no way to tell whether a node sends us a crap block that doesn't belong in the commitment and thus fucks up the whole epoch. What if we request epoch by epoch from different neighbors?